src/main/java/bio/terra/common/retry/SamRetry.java

```diff
@@ -0,0 +1,128 @@
+import org.broadinstitute.dsde.workbench.client.sam.ApiException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+
+/**
+ * SamRetry encapsulates logic needed for retrying Sam API calls. All constants are hard-coded, as
+ * Sam failures are generally just short periods of down-time.
+ *
+ * <p>SamRetry throws either the underlying ApiException from Sam or an InterruptedException.
+ */
+public class SamRetry {
+  private static Logger logger = LoggerFactory.getLogger(SamRetry.class);
+
+  // The retry function starts with INITIAL_WAIT between retries, and doubles that until it
+  // reaches MAXIMUM_WAIT, after which all retries are MAXIMUM_WAIT apart.
+  private static final Duration MAXIMUM_WAIT = Duration.ofSeconds(30);
+  private static final Duration INITIAL_WAIT = Duration.ofSeconds(10);
+  private static final Duration OPERATION_TIMEOUT = Duration.ofSeconds(300);
+  private final Instant operationTimeout;
+
+  // How long to wait between retries.
+  private Duration retryDuration;
+
+  private static final ObjectMapper objectMapper = new ObjectMapper();
+
+  SamRetry() {
+    this.operationTimeout = now().plus(OPERATION_TIMEOUT);
+    this.retryDuration = INITIAL_WAIT;
+  }
+
+  protected SamRetry(Duration timeout) {
+    this.operationTimeout = now().plus(timeout);
+    this.retryDuration = INITIAL_WAIT;
+  }
+
+  @FunctionalInterface
+  public interface SamVoidFunction {
+    void apply() throws ApiException, InterruptedException;
+  }
+
+  @FunctionalInterface
+  public interface SamFunction<R> {
+    R apply() throws ApiException, InterruptedException;
+  }
+
+  public static <T> T retry(SamFunction<T> function) throws ApiException, InterruptedException {
+    SamRetry samRetry = new SamRetry();
+    return samRetry.perform(function);
+  }
+
+  public static <T> T retry(SamFunction<T> function, Duration timeout)
+      throws ApiException, InterruptedException {
+    SamRetry samRetry = new SamRetry(timeout);
+    return samRetry.perform(function);
+  }
+
+  public static void retry(SamVoidFunction function) throws ApiException, InterruptedException {
+    SamRetry samRetry = new SamRetry();
+    samRetry.performVoid(function);
+  }
+
+  public static void retry(SamVoidFunction function, Duration timeout)
+      throws ApiException, InterruptedException {
+    SamRetry samRetry = new SamRetry(timeout);
+    samRetry.performVoid(function);
+  }
+
+  private <T> T perform(SamFunction<T> function) throws ApiException, InterruptedException {
+    while (true) {
+      try {
+        return function.apply();
+      } catch (ApiException ex) {
+        if (isRetryable(ex)) {
+          logger.info("SamRetry: caught retry-able exception: ", ex);
+          sleepOrTimeoutBeforeRetrying(ex);
+        } else {
+          throw ex;
+        }
+      } catch (Exception ex) {
+        throw new InternalServerErrorException("Unexpected exception type: " + ex.toString(), ex);
```

Comment:
I'm not sure this is needed. But maybe atleast `catch (RuntimeException` so we don't accidentally lose an InterruptedException?
as-is, `try { return function.apply(); }` is allowed to throw an InterruptedException by the signature on the SamFunction interface, and this catch will catch and rethrow it as a InternalServerErrorException, which we're trying to avoid.

Reply:
yeah I think this is not needed at all

Reply:
If it is not needed, it should be removed. If you leave it, you should change it to `RuntimeException` so it will not catch `InterruptedException`

Reply:
Should be gone now.